contracts/competition/Competition.sol

```diff
@@ -0,0 +1,169 @@
+        Competition storage competition = competitions[competitionIndex];
+
+        require(competition.memberTeams[msg.sender] == address(0), "You can't join multiple teams.");
+        require(competition.teams[leaderAddress].leader != address(0), "The team does not exist.");
+
+        competition.joinRequests[msg.sender] = leaderAddress;
+
+        emit JoinRequestCreated(competitionIndex, msg.sender, leaderAddress);
+    }
+
+    function approveJoinRequest(uint competitionIndex, address memberAddress) external registrationIsOpen(competitionIndex) {
+        Competition storage competition = competitions[competitionIndex];
+
+        require(competition.joinRequests[memberAddress] == msg.sender, "This member did not apply.");
+        require(competition.memberTeams[memberAddress] == address(0), "This member already joined a team.");
+        require(competition.teams[msg.sender].members.length < competition.maxTeamSize, "Team is full.");
+
+        // referralStorage.setTraderReferralCode(memberAddress, teams[msg.sender].referral);
+        competition.teams[msg.sender].members.push(memberAddress);
+        competition.memberTeams[memberAddress] = msg.sender;
+        competition.joinRequests[memberAddress] = address(0);
+
+        emit JoinRequestApproved(competitionIndex, memberAddress, msg.sender);
+    }
+
+    function cancelJoinRequest(uint competitionIndex) external registrationIsOpen(competitionIndex) {
+        competitions[competitionIndex].joinRequests[msg.sender] = address(0);
+        emit JoinRequestCanceled(competitionIndex, msg.sender);
+    }
+
+    function removeMember(uint competitionIndex, address memberAddress) external registrationIsOpen(competitionIndex) {
+        Competition storage competition = competitions[competitionIndex];
+
+        require(competition.memberTeams[memberAddress] == msg.sender, "This member is not in your team");
+
+        for (uint i = 0; i < competition.teams[msg.sender].members.length; i++) {
+            if (competition.teams[msg.sender].members[i] == memberAddress) {
+                delete competition.teams[msg.sender].members[i];
+                break;
+            }
+        }
+
+        competition.memberTeams[memberAddress] = address(0);
+
+        emit MemberRemoved(competitionIndex, msg.sender, memberAddress);
+    }
+
+    function getCompetition(uint index) external view returns (uint, uint, uint) {
+        return (
+            competitions[index].start,
+            competitions[index].end,
+            competitions[index].maxTeamSize
+        );
+    }
+
+    function getTeam(uint competitionIndex, address leaderAddr) external view returns (address, string memory, bytes32) {
+        Team memory team = competitions[competitionIndex].teams[leaderAddr];
+        return (team.leader, team.name, team.referralCode);
+    }
+
+    function getTeamMembers(uint competitionIndex, address leaderAddr, uint start, uint offset) external view returns (address[] memory) {
```

Comment:
can we use `leaderAddress`? for consistency, we don't use `addr` in other contracts